From cfc65a7fd5bb6bacc0fe96962d21ba2578d609e7 Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Thu, 17 Oct 2013 18:21:07 -0700
Subject: [PATCH 342/974] MIPS: paravirt: Implement _machine_halt

This allows us to exit the VM.

Signed-off-by: David Daney <david.daney@cavium.com>
[Original patch taken from Cavium SDK 3.1.2-568]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 arch/mips/paravirt/setup.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/arch/mips/paravirt/setup.c b/arch/mips/paravirt/setup.c
index 75691e2..80c325e 100644
--- a/arch/mips/paravirt/setup.c
+++ b/arch/mips/paravirt/setup.c
@@ -11,6 +11,7 @@
 #include <asm/bootinfo.h>
 #include <asm/mipsregs.h>
 #include <asm/smp-ops.h>
+#include <asm/reboot.h>
 #include <asm/time.h>
 
 extern struct plat_smp_ops paravirt_smp_ops;
@@ -27,6 +28,16 @@ void __init plat_time_init(void)
 	preset_lpj = mips_hpt_frequency / (2 * HZ);
 }
 
+static void plat_halt_this_cpu(void *ignore)
+{
+	hypcall0(1 /* Exit VM. */);
+}
+
+static void plat_halt(void)
+{
+	on_each_cpu(plat_halt_this_cpu, NULL, 0);
+}
+
 /**
  * Early entry point for arch setup
  */
@@ -48,6 +59,7 @@ void __init prom_init(void)
 			strlcat(arcs_cmdline, " ", COMMAND_LINE_SIZE);
 	}
 	register_smp_ops(&paravirt_smp_ops);
+	_machine_halt = plat_halt;
 }
 
 void __init plat_mem_setup(void)
-- 
2.6.2

